lib/OptimizationMOI/src/OptimizationMOI.jl

```diff
@@ -273,75 +62,70 @@ function __map_optimizer_args(prob::OptimizationProblem,
+    if expr.head == :ref
+        @assert length(expr.args) == 2
+        if expr.args[1] == :x
+            return Expr(:ref, :x, MOI.VariableIndex(expr.args[2]))
+        elseif expr.args[1] != :p
+            error("Expected an expression with respect to `x` and `p` only!")
         end
-    end
-    if prob.ub !== nothing
-        @assert eachindex(prob.ub) == Base.OneTo(num_variables)
-        for i in 1:num_variables
-            if prob.ub[i] < Inf
-                MOI.add_constraint(opt_setup, θ[i], MOI.LessThan(prob.ub[i]))
-            end
+    elseif expr.head == :call && expr.args[1] == :getindex
+        @assert length(expr.args) == 3
+        if expr.args[2] == :x
+            return Expr(:ref, :x, MOI.VariableIndex(expr.args[3]))
+        elseif expr.args[2] != :p
+            error("Expected an expression with respect to `x` and `p` only!")
         end
     end
-    if MOI.supports(opt_setup, MOI.VariablePrimalStart(), MOI.VariableIndex)
-        @assert eachindex(prob.u0) == Base.OneTo(num_variables)
-        for i in 1:num_variables
-            MOI.set(opt_setup, MOI.VariablePrimalStart(), θ[i], prob.u0[i])
+    for i in 1:length(expr.args)
+        expr.args[i] = _replace_variable_indices!(expr.args[i])
+    end
+    return expr
+end
+
+_replace_parameter_indices!(expr, p) = expr
+function _replace_parameter_indices!(expr::Expr, p)
+    if expr.head == :ref
+        @assert length(expr.args) == 2
+        if expr.args[1] == :p
+            return p[expr.args[2]]
+        elseif expr.args[1] != :x
+            error("Expected an expression with respect to `x` and `p` only!")
+        end
+    elseif expr.head == :call && expr.args[1] == :getindex
+        @assert length(expr.args) == 3
+        if expr.args[2] == :p
+            return p[expr.args[3]]
+        elseif expr.args[2] != :x
+            error("Expected an expression with respect to `x` and `p` only!")
         end
     end
-    MOI.set(opt_setup,
-            MOI.ObjectiveSense(),
-            prob.sense === Optimization.MaxSense ? MOI.MAX_SENSE : MOI.MIN_SENSE)
-    if prob.lcons === nothing
-        @assert prob.ucons === nothing
-        con_bounds = MOI.NLPBoundsPair[]
-    else
-        @assert prob.ucons !== nothing
-        con_bounds = MOI.NLPBoundsPair.(prob.lcons, prob.ucons)
+    for i in 1:length(expr.args)
+        expr.args[i] = _replace_parameter_indices!(expr.args[i], p)
     end
-    MOI.set(opt_setup,
-            MOI.NLPBlock(),
-            MOI.NLPBlockData(con_bounds, MOIOptimizationProblem(prob), true))
-    MOI.optimize!(opt_setup)
-    if MOI.get(opt_setup, MOI.ResultCount()) >= 1
-        minimizer = MOI.get(opt_setup, MOI.VariablePrimal(), θ)
-        minimum = MOI.get(opt_setup, MOI.ObjectiveValue())
-        opt_ret = Symbol(string(MOI.get(opt_setup, MOI.TerminationStatus())))
+    return expr
+end
+
+include("nlp.jl")
+include("moi.jl")
+
+function SciMLBase.solve(prob::OptimizationProblem,
```

Comment:
How to do this correctly? Because this now skips: https://github.com/SciML/SciMLBase.jl/blob/master/src/solve.jl#L78

Reply:
`function SciMLBase.__solve(prob::OptimizationProblem,`, you should only be overloading the `__` versions and the direct CommonSolve versions should be handled in the highest level

Reply:
Yes, but where is the `__init` called and how does a user get access to it?
`SciMLBase\src\solve.jl:78` overloads the `CommonSolve` implementation.

```julia
# in SciMLBase
function solve(prob::OptimizationProblem, alg, args...; kwargs...)
    # checks
    __solve(prob, alg, args...; kwargs...)
end

# in OptimizationMOI
function SciMLBase.__solve(prob::OptimizationProblem, alg::SomeAlg, args...; kwargs...)
    cache = __init(prob, alg, args...; kwargs...)
    ___solve(cache)
end

function ___solve(cache)

end
```

But as a user I want:

```julia
prob = OptimizationProblem()
sol = solve(prob, alg)

# or

cache = init(prob, alg)
sol = solve!(cache)
```

So do I have to overload `init` and `solve!`?

Reply:
SciMLBase needs to define the init dispatch with the error checking and dropdown.

Reply:
https://github.com/SciML/SciMLBase.jl/pull/294